signer/src/utxo.rs

```diff
@@ -0,0 +1,964 @@
+    /// number of requests in the transaction.
+    fn compute_request_fee(tx: &Transaction, fee_rate: u64) -> u64 {
+        let tx_fee = tx.vsize() as u64 * fee_rate;
+        let num_requests = (tx.input.len() + tx.output.len()).saturating_sub(2) as u64;
+        tx_fee.div_ceil(num_requests)
+    }
+
+    /// Compute the final amount for the signers' UTXO given the current
+    /// UTXO amount and the incoming requests.
+    ///
+    /// This amount does not take into account fees.
+    fn compute_signer_amount(reqs: &[Request], state: &SignerBtcState) -> Result<u64, Error> {
+        let amount = reqs
+            .iter()
+            .fold(state.utxo.amount as i64, |amount, req| match req {
+                Request::Deposit(req) => amount + req.amount as i64,
+                Request::Withdrawal(req) => amount - req.amount as i64,
+            });
+
+        // This should never happen
+        if amount < 0 {
+            tracing::error!("Transaction deposits greater than the inputs!");
+            return Err(Error::InvalidAmount(amount));
+        }
+
+        Ok(amount as u64)
+    }
+
+    /// Adjust the amounts for each output given the fee.
+    ///
+    /// This function adjusts each output by the given fee amount. The
+    /// signers' UTXOs amount absorbs the fee on-chain that the depositors
+    /// are supposed to pay. This amount must be accounted for when
+    /// minting sBTC.
+    fn adjust_amounts(tx: &mut Transaction, fee: u64) {
+        // Since the first input and first output correspond to the signers'
+        // UTXOs, we subtract them when computing the number of requests.
+        let num_requests = (tx.input.len() + tx.output.len()).saturating_sub(2) as u64;
+        // This is a bizarre case that should never happen.
+        if num_requests == 0 {
+            tracing::warn!("No deposit or withdrawal related inputs in the transaction");
+            return;
+        }
+
+        // The first output is the signer's UTXO. To determine the correct
+        // amount for this UTXO deduct the fee payable by the depositors
+        // from the currently set amount. This deduction is reflected in
+        // the amount of sBTC minted to each depositor.
+        if let Some(utxo_out) = tx.output.first_mut() {
+            let deposit_fees = fee * (tx.input.len() - 1) as u64;
+            let signers_amount = utxo_out.value.to_sat().saturating_sub(deposit_fees);
+            utxo_out.value = Amount::from_sat(signers_amount);
+        }
+        // We now update the remaining withdrawal amounts to account for fees.
+        tx.output.iter_mut().skip(1).for_each(|tx_out| {
+            tx_out.value = Amount::from_sat(tx_out.value.to_sat().saturating_sub(fee));
+        });
+    }
+
+    /// Helper function for generating dummy Schnorr signatures.
+    fn generate_dummy_signature() -> Signature {
+        let key_pair = Keypair::new_global(&mut secp256k1::rand::thread_rng());
```

Comment:
Nit: We might as well include `rand` as an explicit dependency since I'm sure we'll use it more times than just here.

Reply:
Also, we've defaulted to using OsRng in many places in the nakamoto signer already. I don't have strong opinions on either, but it's nice to be consistent.

Reply:
Cool beans, will update.